### ISSUE TITLE
Support for Yarn Plug 'n Play

### ISSUE THREAD

**Issue by empyrical** [react-native-community/cli#27]
This is a meta issue to keep track of the things that need to be tweaked to give RN-CLI support for Yarn Plug 'n Play. This is probably non-exhaustive, more things might be found as others are tweaked.

 - [ ] Global CLI: Support for finding `react-native` and `react-native-local-cli` when PnP is enabled (in progress: #26 )

 - [ ] `react-native bundle`: Support for bundling dependencies with PnP. (Metro team is working on open sourcing its PnP support: facebook/metro#308)

 - [ ] `findPlugins`: Plugin discovery code will need to be adapted to support PnP

 - [ ] Xcode `pbxproj` files all have paths hardcoded in like this:
    > `path = "../node_modules/react-native/React/React.xcodeproj";`

    There is, as far as I know, no way to have this path be dynamic. Ideally there would be some kind of way to have Xcode call a subprocess (like `node resolve-dependency.js` or something) and use it as the path's prefix, but I don't think there's a way. So I think a script might need to be made to refresh these paths and keep them up to date.

    This is also relevant for user libraries and `react-native link` for iOS dependencies

 - [ ] Gradle scripts will need to be updated along with `react-native link`'s Gradle output. Here's an example bit out output for `react-native-webview`:

    >```groovy
    > project(':react-native-webview').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-webview/android')
    > ```

    The path to it is hardcoded to look in `node_modules`. I think a new helper function could be written, that would look something like:

    > `npm_dependency('react-native-webview', 'android')`

    Where the first argument is the name of the node module, and the second is a pathname within the node_module. If PnP isn't present, then it will simply output an old-style path to `node_modules`. A similar function could be made for BUCK.

 - Paths to `node_modules/react-native` are hardcoded in these files and will need to be updated. They could all use `util/findReactNativePath`, potentially.
    - [ ] `local-cli/library/library.js`
    - [ ] `local-cli/eject/eject.js`
    - [ ] `local-cli/generator/templates.js`
    - [ ] `local-cli/link/__tests__/ios/getHeaderSearchPath.spec.js`
    - [ ] `local-cli/upgrade/upgrade.js`
    - [ ] `local-cli/util/findReactNativeScripts.js`

**Comment by oshimayoan** [comment]
I saw #26 was closed. So how is this going?

**Comment by ManAnRuck** [comment]
some news for PnP support?

**Comment by thymikee** [comment]
None yet, unfortunately

**Comment by sebastian-nowak** [comment]
Any estimates on this, since yarn 2 is now stable and officially released? It defaults to PnP.

**Comment by dalinarkholin** [comment]
@arcanis @empyrical @thymikee I realize its been a couple years since this issue was created. Looking through the react-native code base and it is pretty clear steps haven't really been taken by anyone to make this happen. I would like to start a conversation around supporting PNP for react-native, and possible solutions to make that happen. I think all the tasks @empyrical pointed out a couple years ago still apply.

There is so much tied to the existing node_module structure, I am not sure it makes sense to wholesale try and get expo itself to support pnp. Maybe step one is to provide a way for an ejected React native code base to use pnp more cleanly?  

Note to anyone coming across this: I would eject before even trying to use pnp with react-native.

I have dug into this a bit and got things working with yarn berry and pnp enabled. You can see a working [example here](https://github.com/dalinarkholin/react-native-pnp-enabled) 

Note: This is only working for IOS as of right now.

I ran into a lot of issues getting this setup and working but here are the things that I believe have to happen in order for it to work for IOS(this list is likely missing a few things as well as include things that don't actually need to happen, because my memory isn't perfect, and there is a lot going on here) If you pull down my repo, the readme should get things running. (you can see "foobar" getting pulled from a sibling workspace which is cool).

- Add package.json scripts to workspace  for react-native related scripts (yarn pod:install, yarn pod:degenerate, etc). This is needed because of how yarn berry works.
- unplugs
```
"@react-native-community/cli": {
      "unplugged": true
    },
    "@react-native-community/cli-platform-ios": {
      "unplugged": true
    },
    "expo": {
      "unplugged": true
    },
    "expo-cli": {
      "unplugged": true
    },
    "jest-haste-map": {
      "unplugged": true
    },
    "react-native": {
      "unplugged": true
    },
    "react-native-unimodules": {
      "unplugged": true
    }
```

I believe you have to unplug at least the ones where pods are pulled from and are referenced from the PodFile. The PodFile doesn't seem to be able to reference files within a cached .zip.



- modify the paths in Podfile to point at the unplugs(this isn't great since those hash's can change after a yarn install at times. Script?
- Not sure autolinking works at all, so I manually added the one missing that I needed.
- Add a metro.config.js as seen [here](https://github.com/dalinarkholin/react-native-pnp-enabled/blob/master/packages/react-native-pnp-example/metro.config.js). This added a pnpResolver which it was included with Metro at one point, but seems to have been removed. The code here is pretty darn close to what was included before.

PNP Enabled issues.
______
- I got a chain of issues after adding pnp resolver to metro. First error was [this guy](https://github.com/facebook/metro/issues/330) . So I added a patch to metro see .patches/metro.patch in repo.
- Second was that crypto is not a npm module, so i created a passthrough workspace(see repo) to point at react-native-crypto. (learned about resolutions after the fact, probably could just do something like "crypto: react-native-crypto" Not sure that works.)
- Third issue was that `process.version` was undefined which crypto looks at from a submodule pbkdf2. Added a patch for it, but it did just get merged as well here: https://github.com/crypto-browserify/pbkdf2/pull/85#event-3312770971 it might be a better solution here to actually include process? not sure how this works by default.
- If you checkout the .yarnrc.yaml you will see an insanely long list of packageExtensions. That it definitely not 100% accurate. Pretty sure some of those should be peerDependenicies, but they got me moving forward, and are likely needed in some form.

I ran into a million other issues, but I believe they were issues from trying to eject with pnp enabled initially.

Thoughts on any of this? If nothing comes of this, at at least there is an example of pnp working with react native on the webs.

To wrap up, IMO, yarn V2 is the future of JS/TS development. It was sort of crazy how many open and closed issues I ran into digging into this within the RN ecosystem...most of them dependency issues during buildtime, runtime, etc...all of which just sort of magically work, until it doesn't. As a community we need to start moving forward with Yarn V2, where we have a better grip of what is going on with our dependencies.

**Comment by natew** [comment]
@dalinarkholin Just wanted to say thank you. I wonder why this has been dead for so long, it's really odd to see two of the biggest FB projects totally out of sync for... going on 3 years.

Anyway, hope to see this move!

**Comment by marudy** [comment]
@dalinarkholin thanks for that. Wanted to try your example to verify that indeed I could build of course with yarn berry and also have a look at `yarn install` performance because this is the main reason for me wanting to move out from yarn v1. 

After this was successful , I took some time to configure yarn berry into my project and see `yarn install` performance. The results speak for themselves:

**yarn v1**
- yarn install with no `node_modules` folder: 5 mins, 46 secs
✨  Done in 44.54s. ✨  Done in 301.41s.

**yarn berry with PnP**
- without cached install and without unplugged files, let's call this first time install: 1 min, 46secs
➤ YN0000: Done with warnings in 1m 46s

- without cached install files: 45 seconds
➤ YN0000: Done with warnings in 45s 587ms

- with zero installs and cache: 4 seconds
➤ YN0000: Done with warnings in 3s 808ms

I am working on a project with huge amount of packages and dependencies and every time I want to do a fresh install it takes a lot... Heck my CI has to eat up that time as well. I think it only makes sense move towards that direction. Also I believe there are lots of us willing to do dev tasks to complete this. I would be happy to help with parts where my knowledge and expertise makes sense if we can put this down to concrete tasks.

@arcanis @empyrical @thymikee what do you guys think? Shall we give berry a chance ?

**Comment by thymikee** [comment]
@marudy we're happy with accepting any help on how to make Berry work with React Native. Even in a form of the documentation :)

**Comment by jgornick** [comment]
Hi @thymikee and @marudy, it's been a few months since this was last commented on. I wanted to check-in to see if there has been any progress made with regards to React Native compatibility with PnP? This could be code changes, patches, documentation, workarounds, blog posts, anything 😄 

We're looking to use PnP in a monorepo that also has a React Native application and wanted to check-in before we start digging.

Thanks!

**Comment by thymikee** [comment]
No progress unfortunately

**Comment by ariccio** [comment]
*whimsically shakes fist*

Darn it, I'm trying to switch to the fancy new yarn to make my monorepo all nice and clean, but the fact that this isn't just "batteries included" yet is a real pain point. Anybody know what's going on with support?

**Comment by PrimeDominus** [comment]
yarn berry please please please pretty please

**Comment by airtonix** [comment]
noone at facebook cares is the simple truth - we'd see progress on this if they did.

**Comment by tj-mc** [comment]
Yarn Berry currently works fine so long as you use the `nodeLinker: node-modules` option in your `.yarnrc.yml`. This allows you to check in your `.yarn/cache`, then yarn creates the `node_modules` folder when you run `yarn install`. Because there's no network fetching, the install step will last a few seconds at most. This way you get most of the advantages of PnP until it is fully supported. I'll share a blog post when its published,

**Comment by RobinBobin** [comment]
I didn't see `.yarn/cache` in `git status -u`, but then I saw I was just ignoring it by following [the guide](https://yarnpkg.com/getting-started/qa#which-files-should-be-gitignored) too strictly. Maybe it will help someone...

**Comment by alamothe** [comment]
@tj-mc I disagree brother. The `node-modules` is completely broken, especially if you work with TypeScript. We have dozens of shared workspaces in our monorepo and there is no way we would revert from PNP because of lack of support in RN.

**Comment by tj-mc** [comment]
@alamothe I haven't experienced anything odd using this in a Typescript repo in a single workspace.

**Comment by michaelfaith** [comment]
Curious if this is going to be prioritized at some point?

**Comment by fguitton** [comment]
Not stale 🎈

**Comment by Slapbox** [comment]
Is there any work ongoing to address this? I used to think this was going to happen eventually, but now I think it might not even happen if Yarn v1 had a CVSS 10 rated bug.

**Comment by michaelfaith** [comment]
> Also keep in mind Yarn 1 is not deprecated and is still actively worked on. 

Not really though.  Yes, they'd probably fix that CVSS 10 rated bug, but they're generally not fixing bugs or doing any enhancements that aren't related to yarn berry.  There are a couple bug fixes that have impacted my group that they've flat out said they're not going to fix in yarn 1, as they've already fixed it in berry.  

> Yarn Berry actually works fine with RN, you just need to use nodeLinker: node-modules

I feel like the node-modules linker option is more of a stopgap to help people transition to PNP, not to be a final destination.  This was especially useful as various libraries took longer to provide support.  Ultimately the goal is to end up with PNP, hence this issue.

**Comment by arcanis** [comment]
> I feel like the node-modules linker option is more of a stopgap to help people transition to PNP, not to be a final destination.

Just to clarify: the node_modules linker in modern releases of Yarn isn't a stopgap: it's a really good implementation (so good in fact that even pnpm relies on it [as a dependency](https://github.com/pnpm/pnpm/blob/634d6874b75eec2e149de036d7fc7efa7afb7738/pkg-manager/real-hoist/package.json#L36)), significantly improved other the one in Yarn 1 (a couple of hoisting issues were fixed, new features were added), that we don't plan to sunset.

With that being said, it's true that a significant chunk of our users prefer to use PnP, so I agree it'd be nice to see this issue solved someday.

**Comment by michaelfaith** [comment]
Good to know.  I just assumed that was implemented as a means to help people move towards PNP.  Thanks for clarifying. (PNP is really nice though haha)

**Comment by levino** [comment]
Does anyone here actually work at Meta and can make any kind of reliable comment on this? I think "trying to predict what Meta might or might not do about this" is not really leading anywhere. I think it is reasonable to prevent this issue from getting closed as stale and hope that Meta eventually starts hiring again and then has some excess money to spend on dev time for RN. Until then probably nothing else can be done.

**Comment by MohammadKurjieh** [comment]
Any updates on this?

**Comment by alamothe** [comment]
Meta only cares about profit and stock price at this point, open source is not their priority.

**Comment by danielcampo** [comment]
Any place to look to possibly get started on support for this ourselves or is what's needed blocked on the Meta side?

I see there was a PR submitted some time ago, but it seems they hit a wall: https://github.com/goorea/muscleman/issues/96#issuecomment-997160566

**Comment by robhogan** [comment]
> Does anyone here actually work at Meta and can make any kind of reliable comment on this? I think "trying to predict what Meta might or might not do about this" is not really leading anywhere.

That's fair. We are rolling out support for symlink-based `node_modules` layouts, and with that support for most other modern packagers and package layouts, but  we have no current plans to support PnP. 

That would take non-trivial work in Metro's resolver and transformer, and there may be implications for native builds too (eg, native builds sometimes write build intermediates inside `node_modules`, which would be an especially bad idea in PnP). 

The React DevX team isn't huge, we have to prioritise, and there's lots we're working on that we think will bring more benefit to RN devs - new debugger, improved dev tools, faster and more reliable bundling. We hope folks will be able to make use of improved packager support in the latest versions, just not PnP for now.

**Comment by szymonrybczak** [comment]
As @robhogan said, we have no plans for supporting PnP, so I'm closing this issue. We know that package managers are important and we'll try to improve this space in future releases.

**Comment by Slapbox** [comment]
@szymonrybczak can you clarify what exactly this might mean?

> We know that package managers are important and we'll try to improve this space in future releases.

It's hard to parse the meaning when the preceding line says there's no plans to support PnP.

**Comment by levino** [comment]
In my experience bun is even better than yarn with zero installs. Fwiw this topic has become kinda obsolete for me.

**Comment by OneHoopyFrood** [comment]
> As @robhogan said, we have no plans for supporting PnP, so I'm closing this issue. We know that package managers are important and we'll try to improve this space in future releases.

It would seem to me that the meta-issue should remain open as it sounds like there are tentative plans to support this in the future, if not concrete ones.

**Comment by Nantris** [comment]
I agree this should not be closed. The closing comment basically says, "We have no immediate plans to fix it, but someday we'll try to fix it because we know it's important."

I don't see why the issue should be closed with that in mind.

**Comment by Nantris** [comment]
This issue continues to kill me with regards to disk space. My 200mb app requires, somehow, 10gb of disk space to build and it sure would be a help if I could minimize the room my `node_modules` was taking up.

**Comment by Nantris** [comment]
PnP would spare me the space that the largely duplicated `node_modules` folders of projects that share common technologies take up. Sure, I can run around like a maniac deleting and re-creating `node_modules` in a bunch of different projects, but that's extremely tedious and probably not helping the SSD's longevity.

**Comment by levino** [comment]
As far as I understand it, with "zero install" you commit all the sources for all the deps to the repository. So there is **only** a local cache, no global one. The feature is nice. `yarn install` takes maybe 10 seconds or so after a clone. But I heavily doubt that you will save disk space or disk writes by using `yarn` pnp.

**Comment by BasixKOR** [comment]
Yarn can and does use global cache alongside local cache and Zero-install is an opt-in feature. Note the PnP and Zero-install aren't the same thing.

And importantly, This is not a place to discuss the Yarn's merits, please don't spam the people who subscribed to this issue.
